src/hffs/fs.py

```diff
@@ -88,58 +86,84 @@ class HfFileSystem(fsspec.AbstractFileSystem):
     ```python
     >>> import fsspec
 
-    >>> # Read a remote file
-    >>> with fsspec.open("hf://datasets/my-username/my-dataset:/remote/file/in/repo.bin") as f:
+    >>> # Read/write files
+    >>> with fsspec.open("hf://my-username/my-model/pytorch_model.bin") as f:
     ...     data = f.read()
-
-    >>> # Write a remote file
-    >>> with fsspec.open("hf://datasets/my-username/my-dataset:/remote/file/in/repo.bin", "wb") as f:
+    >>> with fsspec.open("hf://my-username/my-model/pytorch_model.bin", "wb") as f:
     ...     f.write(data)
     ```
     """
 
     root_marker = ""
     protocol = "hf"
 
-    @huggingface_hub.utils.validate_hf_hub_args
     def __init__(
         self,
-        repo_id: str,
+        *args,
         token: Optional[str] = None,
-        repo_type: Optional[str] = None,
-        revision: Optional[str] = None,
         endpoint: Optional[str] = None,
-        **kwargs,
+        revision: Optional[str] = None,
+        **storage_options,
     ):
-        super().__init__(self, **kwargs)
-
-        if repo_type not in huggingface_hub.constants.REPO_TYPES:
-            raise ValueError(f"Invalid repo type, must be one of {huggingface_hub.constants.REPO_TYPES}")
-
-        self.repo_id = repo_id
-        self.endpoint = endpoint if endpoint is not None else huggingface_hub.constants.ENDPOINT
-        self.token = token if token is not None else huggingface_hub.HfFolder.get_token()
-        self.repo_type = repo_type if repo_type is not None else huggingface_hub.constants.REPO_TYPE_MODEL
+        super().__init__(*args, **storage_options)
+        self.token = token
+        self.endpoint = endpoint
         self.revision = revision
         self._api = huggingface_hub.HfApi(
             endpoint=endpoint, token=token, library_name="hffs", library_version=__version__
         )
 
-    def _dircache_from_repo_info(self):
-        repo_info = self._api.repo_info(
-            self.repo_id, revision=self.revision, repo_type=self.repo_type, token=self.token, files_metadata=True
-        )
+    def _resolve_repo_id(self, path: str) -> Tuple[str, str, str]:
+        path = self._strip_protocol(path)
+        if path.split("/")[0] in huggingface_hub.constants.REPO_TYPES_MAPPING:
+            if "/" not in path:
+                raise NotImplementedError("Listing repositories is not implemented.")
+            repo_type, path = path.split("/", 1)
+            repo_type = huggingface_hub.constants.REPO_TYPES_MAPPING[repo_type]
+        else:
+            repo_type = None
+        if path.count("/") > 0:
+            try:
+                namespace, repo_name, *parts_in_repo = path.split("/")
+                repo_id_with_namespace = f"{namespace}/{repo_name}"
+                self._api.repo_info(repo_id_with_namespace, repo_type=repo_type)
```

Comment:
I think it would make sense to move the "try/except + `repo_info`" in a separate helper (`repo_exists()`?). Since it will be called multiple times and potentially on the same repo (for example in `cp`), I think it can make sense to wrap `repo_exists` in a `@lru_cache` decorator. Canonical models and datasets almost never change so it shouldn't be a problem. If a repo is deleted and the entry is still in the cache,  the subsequent action will raise a RepoNotFound error anyway (worse case scenario if we are really worried about caching issues, we can always have a flag to disable it but I'm not sure it's useful)

Reply:
When using @lru_cache you can also clear the cache manually (with `my_function.clear_cache()`) if you want. But a dictionary should be fine as well if we consider users will not run it on thousands on repos :D